### ISSUE TITLE
After Deleting TLSProfile and VirtualServer CRD it still left a *_443_tls_irule on the big-ip

### ISSUE THREAD

**Issue by mitchellmaler** [F5Networks/k8s-bigip-ctlr#1739]
#### Setup Details
 CIS Version : 2.3.0
 Build: f5networks/k8s-bigip-ctlr:2.3.0       
 BIGIP Version: BIG-IP 15.1.2 Build 0.0.9 Final 
 AS3 Version: 3.26.0
 Agent Mode: AS3
 Orchestration: K8S
 Orchestration Version:  v1.18.10
 Pool Mode: Nodeport
 Additional Setup details: <Platform/CNI Plugins/ cluster nodes/ etc>

#### Description  

I have two CIS controllers running in CRD mode each pointed to a single device that is in a pair. I create a VirtualService and TLSProfile which in turn creates the VirtualServer, Policy, iRule, Monitor, and referenced an existing TLS Profile on the Big-IP. I then delete the two CRD resources and I see all the objects deleted on the Big-IP except for the vsname_443_tls_irule.
  

#### Steps To Reproduce
1) Create VS and TLSProfile CRD (referencing existing big-ip profile)
2) Delete created VS and TLSProfile CRD
3) Look under irules in the partition and that virtualservers/tlsprofile irule will still exist

#### Expected Result 

All created resources in the big-ip are deleted that are related to the TLSProfile and VirtualService deleted.

#### Actual Result  

The vsname_443_tls_irule is not deleted from the big-ip.

**Comment by trinaths** [comment]
@mitchellmaler unable to reproduce this issue. Refresh the GUI and check the data. the deleted VS and TS profile are deleted and no stale config from CIS exists with BIGIP. Hope this helps. closing this issue.

**Comment by mitchellmaler** [comment]
@trinaths Before logging this I have recreated it multiple times and even refreshing the objects on both big-ip devices in my HA Pair. It is reproducible on my side which means there is some way this issues occurs.


Here is my evidence:
I create a test_k8s VS and TLS Profile and all objects are created in the Big-IP


![Screen Shot 2021-03-09 at 10 53 52 AM](https://user-images.githubusercontent.com/4107699/110507942-20f8fa80-80c6-11eb-8364-53bc63be6b8a.png)
![Screen Shot 2021-03-09 at 10 54 05 AM](https://user-images.githubusercontent.com/4107699/110507944-21919100-80c6-11eb-85c3-2e46e47fe293.png)
![Screen Shot 2021-03-09 at 10 54 24 AM](https://user-images.githubusercontent.com/4107699/110507946-21919100-80c6-11eb-9d09-bfc0621410a7.png)
![Screen Shot 2021-03-09 at 10 54 33 AM](https://user-images.githubusercontent.com/4107699/110507947-21919100-80c6-11eb-90e3-636502236965.png)
![Screen Shot 2021-03-09 at 10 54 46 AM](https://user-images.githubusercontent.com/4107699/110507950-222a2780-80c6-11eb-90cd-3b0462471b8e.png)

Then I delete the two resources:
![Screen Shot 2021-03-09 at 10 57 20 AM](https://user-images.githubusercontent.com/4107699/110508057-3c640580-80c6-11eb-90df-0c2f8aa2c200.png)

Which results in the irule still existing but nothing else (other than my manually created vs and pool):
![Screen Shot 2021-03-09 at 10 58 05 AM](https://user-images.githubusercontent.com/4107699/110508410-9795f800-80c6-11eb-8b8d-b4181b5f589a.png)
![Screen Shot 2021-03-09 at 10 58 16 AM](https://user-images.githubusercontent.com/4107699/110508411-982e8e80-80c6-11eb-966a-d3c5476a7e26.png)
![Screen Shot 2021-03-09 at 10 59 20 AM](https://user-images.githubusercontent.com/4107699/110508412-982e8e80-80c6-11eb-82b7-416b0264dfba.png)
![Screen Shot 2021-03-09 at 10 59 25 AM](https://user-images.githubusercontent.com/4107699/110508413-98c72500-80c6-11eb-896f-bde4b9abdacd.png)

**Comment by mitchellmaler** [comment]
@trinaths @mdditt2000 Is there any other information I can give that might help track down why this irule is orphaned? We are testing out the migration to the CIS ctrl v2 using CRD and just making sure everything is working as expected. In this case we don't want orphaned irules if we delete the resources from our k8s cluster.

**Comment by trinaths** [comment]
@mitchellmaler - in the above, VS "test" still exists. is the iRule attached to that VS ? When I test this, I see no stale config left in BIGIP.

**Comment by trinaths** [comment]
@mitchellmaler - Can you manually delete the irule and retry that again ?

**Comment by mitchellmaler** [comment]
I just tested again. This time even deleting that test vs and test pool that was in common to make sure the device was empty. The resources are all created. When I delete them both only one out of the three irules are deleted.

## Before Creation:
![Screen Shot 2021-03-09 at 1 36 35 PM](https://user-images.githubusercontent.com/4107699/110527462-9bcd1000-80dc-11eb-9c53-b481c41efab8.png)
![Screen Shot 2021-03-09 at 1 36 49 PM](https://user-images.githubusercontent.com/4107699/110527459-9b347980-80dc-11eb-8343-81b415f51fbf.png)
![Screen Shot 2021-03-09 at 1 36 42 PM](https://user-images.githubusercontent.com/4107699/110527452-9a034c80-80dc-11eb-822c-e88bc4fb7307.png)
![Screen Shot 2021-03-09 at 1 36 58 PM](https://user-images.githubusercontent.com/4107699/110527461-9b347980-80dc-11eb-846d-f92f6df3f5e5.png)

## Resource Creation using CRD (VS and Profile):
![Screen Shot 2021-03-09 at 1 38 34 PM](https://user-images.githubusercontent.com/4107699/110527580-c3bc7380-80dc-11eb-9e18-936d4a53e055.png)
![Screen Shot 2021-03-09 at 1 39 06 PM](https://user-images.githubusercontent.com/4107699/110527700-e9497d00-80dc-11eb-920e-290f8f9da3e8.png)
![Screen Shot 2021-03-09 at 1 39 13 PM](https://user-images.githubusercontent.com/4107699/110527703-e9e21380-80dc-11eb-9cb9-d6bdc42b756e.png)
![Screen Shot 2021-03-09 at 1 39 21 PM](https://user-images.githubusercontent.com/4107699/110527706-e9e21380-80dc-11eb-83eb-d323ab1fe462.png)
![Screen Shot 2021-03-09 at 1 39 28 PM](https://user-images.githubusercontent.com/4107699/110527709-ea7aaa00-80dc-11eb-97ea-24a9073b0b7e.png)

## Delete CRD (VS and Profile) and wait a while/hitting refresh:
![Screen Shot 2021-03-09 at 1 40 33 PM](https://user-images.githubusercontent.com/4107699/110528409-bce23080-80dd-11eb-962d-fa95adc85e3f.png)
![Screen Shot 2021-03-09 at 1 45 18 PM](https://user-images.githubusercontent.com/4107699/110528399-bb186d00-80dd-11eb-9cf5-1be1c0228ab3.png)
![Screen Shot 2021-03-09 at 1 45 24 PM](https://user-images.githubusercontent.com/4107699/110528403-bc499a00-80dd-11eb-9e79-a3744e754173.png)
![Screen Shot 2021-03-09 at 1 40 56 PM](https://user-images.githubusercontent.com/4107699/110528406-bc499a00-80dd-11eb-94cb-5d011c29d687.png)
![Screen Shot 2021-03-09 at 1 40 46 PM](https://user-images.githubusercontent.com/4107699/110528408-bce23080-80dd-11eb-95cd-016844928cba.png)

**Comment by mitchellmaler** [comment]
@trinaths I made sure to wait a while and hit refresh a bunch.

**Comment by mdditt2000** [comment]
@trinaths i cannot reproduce this issue

My CRD 

https://github.com/mdditt2000/kubernetes-1-19/blob/master/cis%202.3/github/securework/crd/vs-securework.yaml

I see the irules correctly add when creating the CRD

```
[kube@k8s-1-19-master crd]$ kubectl create -f vs-securework.yaml
virtualserver.cis.f5.com/securework-virtual created
```
![image](https://user-images.githubusercontent.com/12733938/110545835-e9507980-80ea-11eb-9da6-74174e8c38e9.png)

Added correctly

```
[kube@k8s-1-19-master crd]$ kubectl delete -f vs-securework.yaml
virtualserver.cis.f5.com "securework-virtual" deleted
```
Gone

![image](https://user-images.githubusercontent.com/12733938/110545986-1f8df900-80eb-11eb-8ad9-461be44c95d2.png)

@mitchellmaler does your crd look the same?

**Comment by mitchellmaler** [comment]
@mdditt2000 I disabled device-group sync (automatic) to make sure that wasn't causing a conflict but I am still seeing the 443_tls_irule after deleting the VirtualServer and TLSProfile CRD. I also redeployed the controller this time with 
```
  manage_configmaps: false
  manage_ingress: false
  custom_resource_mode: true
  agent: as3
```
basically making sure only as3 and crd mode was on even though I am pretty sure deploying with CRD mode automatically does this.

Here is my VirtualServer and TLSProfile Resource Definition: 

```
apiVersion: "cis.f5.com/v1"
kind: VirtualServer
metadata:
  name: test-k8s
  namespace: istio-system
  labels:
    f5cr: "true"
spec:
  virtualServerAddress: "REDACTED"
  httpTraffic: redirect
  tlsProfileName: test-k8s
  iRules:
    - /Common/X-Forwarded-Proto
    - /Common/X-Request-Start
    - /Common/Strip_Server_Name
  pools:
  - path: /
    service: istio-ingressgateway
    servicePort: 4043
    monitor:
      type: http
      send: "GET / HTTP/1.1\r\nHost: \r\nConnection: Close\r\n\r\n"
      # revc:
      interval: 3
      timeout:  5
```

**Comment by mdditt2000** [comment]
@mitchellmaler do you have 30 min this afternoon for a zoom call?

**Comment by mitchellmaler** [comment]
@mdditt2000 Sure I am free most of the afternoon.

**Comment by mdditt2000** [comment]
@trinaths I can see the issue. Will share recording later today with the CIS team

**Comment by mitchellmaler** [comment]
Is this fixed by https://github.com/F5Networks/k8s-bigip-ctlr/pull/1753?

**Comment by trinaths** [comment]
@mitchellmaler - Can you reverify this issue. Its fixed with #1753

**Comment by mitchellmaler** [comment]
This was fixed by #1753. We can close this issue.
/close
